### PR TITLE
utils: throw error on malformed input in base64 decode

### PR DESCRIPTION
**Pull request by nuivall**
Several cases where fixed in this patches, all are related to processing of malformed base64 data. Main purpose was to bring alternator implementation closer to what DynamoDB does. We now:
- Throw error when padding is missing during base64 decoding
- Throw error when base64 data is malformed
- In alternator when invalid base64 data is fetched from DB (as opposed to being part of user's request) we now exclude such row during filtering

Additionally some small code quality improvements:
- avoid unnecessary type conversions in calls to rjson:from_strings functions
- avoid some copy constructions in calls to rjson:from_strings functions

Fixes https://github.com/scylladb/scylladb/issues/6487

### CODE REVIEW

**Comment by nyh** [comment]
> In the issue it was discussed whether to allow or disallow missing padding.

In the final PR, please rewrite this first commit (the series' cover letter) to be self contained, and explain what the problem was and how it was fixed without assuming the reader read the issue.

> I went with disallowing it as I think it's better to reject bad data entering the system early (e.g. perhaps padding was missing because buggy user code sent only part of the desired data and trimmed the rest accidentally).
> Also I think compatibility with dynamdb API on unhappy path will make users life easier.

I agree.

> I am setting this PR to draft because I am unsure if I handled errors in fully idiomatic for the project way. Also are we aiming into "__type" compatibility with dynamodb? For instance they return `...#SerializationException` and `...#ValidationException` for missing padding and malformed data, for now I am using single case returning `...#ValidationException`. 

If it's not too difficult or ugly, I would try to return the same error code as DynamoDB does. Usually it's easy (just throw the "right" api_error class instead of the "wrong" one), and makes writing a test that passes on both DynamoDB and Alternator easier. Although I guess it isn't criticial if for some reason it is very hard - or unreasonable (why?) - to use the same error as DynamoDB does.

>Also is re-throwing approach ok?

I don't know what this means. I'll take a look at the code.

> @nyh please advise on the above.
> 
> Note there is also an edge case of handling this in base64_begins_with - I haven't tested this and left the matters as they were. I suspect current codebase will return false, it may be fine for now or we may want to create another issue to check that (probably lower priority).

When we solve an issue, let's try to close all its corner cases because it's unlikely we'll ever want to look at this specific issue again (and we definitely don't want to spend time to debug something we already considered). In this case, if we know base64_begins_with() has a different code base, we should write a test that exercises it. We actually have a C++ unit test specifically for base64_begins_with() alone, but this function is also used for the "begins_with()" condition so you should be able to fairly easily concoct a request that causes it to be used and see what happens.

An important note (and also important to test...): As you can see in the `check_BEGINS_WITH()` code, there is an important difference between a parameter that came from the user (which should cause the request to fail - throwing) and one that came from the database (which should cause the the item to be skipped, "bad = true"). You need to do this in this case too - if we can't allow base64_begins_with() to throw if there's a problem with data in the database and not in the user input. It looks like it won't be very easy to do this cleanly.

File: utils/base64.cc
Comment (nyh) [inline utils/base64.cc:73]: You asked me about how/whether to use exceptions in the base64_decode function.

If you take a look at base64_decode_string(), you can see that today if it finds a problem it just stops decoding and leaves the string partial - or even empty. We have in `db/hints/sync_point.cc` code which assumes that an empty result means fail to parse. First of all, if you change this, you may need to change sync_point.cc too.

Another option is to make base64_decode_string do what sync_point.cc thinks it does - return an empty string on any decode error. The caller than needs to check if the result is empty but the encoded string was not, it means an error. But the caller doesn't know (and probably doesn't care) exactly what kind of encoding error it was.
This approach makes it easier to generate the error at the last moment - e.g., silently skip an item if it's read from the database and throw an exception from the query if the query contained the bad argument.

I don't really know what I prefer. It's your call, I guess.

File: test/alternator/test_manual_requests.py
Comment (nyh) [inline test/alternator/test_manual_requests.py:250]: Oh, now I see why you added this put_wrong_base64_item() function.

File: utils/base64.cc
Comment (nyh) [inline utils/base64.cc:20]: Maybe static const_expr?

**Review by nyh — CHANGES_REQUESTED** [review]

File: test/alternator/test_manual_requests.py
Comment (nuivall) [reply test/alternator/test_manual_requests.py:252]: ok, got it, will rely only on status code here

File: utils/base64.cc
Comment (nuivall) [reply utils/base64.cc:73]: `db/hints/sync_point.cc` shouldn't be the problem as I understand it immediately throws an error, it will be different class, but later this code treats all error classes the same:
```c++
      try {
          sync_point = db::hints::sync_point::decode(encoded);
      } catch (std::exception& e) {
          throw httpd::bad_param_exception(format("Failed to parse the sync point description {}: {}", encoded, e.what()));
      }
```
Although the difference is that if input is empty `sync_point` will additionally throw but alternator will not, this seems to be correct from decoding perspective. And keeps current `sync_point` behavior unchanged.

Because in alternator we have couple usages of this func and we want essentially the same error behavior I'll stay with throwing approach, it reduces the places I need to change.

File: utils/base64.cc
Comment (nuivall) [reply utils/base64.cc:20]: sure

File: test/alternator/test_manual_requests.py
Comment (nuivall) [reply test/alternator/test_manual_requests.py:268]: added

File: test/alternator/test_manual_requests.py
Comment (nuivall) [reply test/alternator/test_manual_requests.py]: added

**Comment by nuivall** [comment]
@nyh v2 ready for review, main change is closing the loophole with expressions you mentioned.

Also note that `test_base64_malformed_cond_expr` demonstrates __correctness__ issue in this special case. To better show this some print may be added:
```diff
         r = scan_with_binary_data_in_cond_expr(dynamodb, test_table_b, exp, exp_attr)
-        assert r.status_code == 400, "Failed on expression \"%s\"" % (exp)
+        print(exp, " -> ", r.text)
+        #assert r.status_code == 400, "Failed on expression \"%s\"" % (exp)
```
ran on master shows this:
```
test_manual_requests.py::test_base64_malformed_cond_expr fixture creating new table alternator_Test_1669738784288
c > :v  ->  {"Count":0,"ScannedCount":1,"Items":[]}
:v > c  ->  {"Count":0,"ScannedCount":1,"Items":[]}
NOT c > :v  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
contains(c, :v)  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
contains(:v, c)  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
c between :v and :v  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
:v between c and c  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
begins_with(c, :v)  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
begins_with(:v, c)  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"QIA5z+1iYdN8gA=="},"c":{"B":"ZmVmZQ=="}}]}
```

also when we change the value of const `:v`:
```
-    malformed_data = "ZmVmZQ=!" # has the same length as c_data to test begins_with
+    malformed_data = "!" # has the same length as c_data to test begins_with
```

results are different but also not correct:

```
c > :v  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"zlHv3Qsi7b+V1Q=="},"c":{"B":"ZmVmZQ=="}}]}
:v > c  ->  {"Count":0,"ScannedCount":1,"Items":[]}
NOT c > :v  ->  {"Count":0,"ScannedCount":1,"Items":[]}
contains(c, :v)  ->  {"Count":1,"ScannedCount":1,"Items":[{"p":{"B":"zlHv3Qsi7b+V1Q=="},"c":{"B":"ZmVmZQ=="}}]}
contains(:v, c)  ->  {"Count":0,"ScannedCount":1,"Items":[]}
c between :v and :v  ->  {"Count":0,"ScannedCount":1,"Items":[]}
:v between c and c  ->  {"Count":0,"ScannedCount":1,"Items":[]}
begins_with(c, :v)  ->  {"Count":0,"ScannedCount":1,"Items":[]}
begins_with(:v, c)  ->  {"Count":0,"ScannedCount":1,"Items":[]}
``` 

This happens because on master base64_decode function just stops decoding on invalid character but doesn't return any error.

**Comment by nuivall** [comment]
@nyh ping, the tests are passing

**Comment by nuivall** [comment]
> I'll review this series now. Please edit the first comment of the pull request - it will become the cover letter of the merge commit when I merge it.

Thank you! Comment changed.

File: alternator/conditions.cc
Comment (nyh) [inline alternator/conditions.cc:237]: Can you catch the specific exception you expect instead of every type?

File: alternator/serialization.cc
Comment (nyh) [inline alternator/serialization.cc:64]: It's scary that unwrap_bytes() looks like it can return an unset optional even if the second parameter is true. But we're safe because it can't. Maybe this needs to be documented in a comment on unwrap_bytes() or here in the call.

File: alternator/serialization.cc
Comment (nyh) [inline alternator/serialization.cc:62]: If I understand correctly, this from_json_visitor code gets called only during **write** operations, so I think it's fine that a write operation fails with an exception if something is wrong in what it tries to write (regardless of where this content comes from). So I think this "true" is correct.

File: alternator/serialization.cc
Comment (nyh) [inline alternator/serialization.cc:329]: I think you can greatly simplify this function by doing here
```
return rjson::base64_decode(value);
```
then you don't need the `bytes b` or the `return std::move(b)`.

File: alternator/serialization.hh
Comment (nyh) [inline alternator/serialization.hh:65]: Actually I see above api_error::serialization. That's not the same thing as ValidationException, right?

File: test/alternator/test_manual_requests.py
Comment (nyh) [inline test/alternator/test_manual_requests.py:301]: Interesting. So maybe we should have a new xfail-ing test that demonstrates this difference? I think it's not important to fix, but maybe it would be good to know.
But not necessary now.

**Review by nyh — CHANGES_REQUESTED** [review]
Looks mostly good, thanks!
I only left a few minor comments.

File: index/secondary_index.cc
Comment (nuivall) [reply index/secondary_index.cc:132]: I forgot how exactly I found it, I think I was just reading how types are converted one into another around rjson code.

File: alternator/serialization.cc
Comment (nuivall) [reply alternator/serialization.cc:329]: yes, that's true

File: alternator/serialization.hh
Comment (nuivall) [reply alternator/serialization.hh:65]: yes, thanks, forgot to update after I made the code to distinguish between those two as dynamo does

File: test/alternator/test_manual_requests.py
Comment (nuivall) [reply test/alternator/test_manual_requests.py:301]: yes, I was thinking, should we have xfail tests for things we don't want to fix? maybe. I think we don't deserialize here as an optimization.

File: alternator/serialization.cc
Comment (nuivall) [reply alternator/serialization.cc:64]: will add to unwrap_bytes docs, it's kind of mentioned but indirectly, should be more clear

File: alternator/conditions.cc
Comment (nuivall) [reply alternator/conditions.cc:237]: yes. Is it because we can also return some kind of out of memory exception or similar indicating that's internal error instead of user error?

File: schema.cc
Comment (nuivall) [reply schema.cc:1763]: I don't think it copies it, sstring constructor does `reinterpret_cast`, and then from_string passes it by const reference. Although perhaps I missed one thing, that it uses strlen while string_view does traits_type::length so that may be more efficient if strlen can't be optimized.

Goal of this particular line change was just to select the right overload explicitly.

Due to abovementioned strlen I will change sstring to string_view.

File: utils/rjson.hh
Comment (nuivall) [reply utils/rjson.hh:192]: > it's just book-keeping in the compiler

I am not sure. It would be great to dump small piece of the code to some kind of IR/SSA or even asm but annotated with c++ code. It's very easy to do in Golang but for C++ I only know how to dump the whole codebase as LLVM bitcode but that's alone not helpful.

What I see during debugging release build is that it enters:
```c++
  operator std::basic_string_view<char_type>() const noexcept {
        ...
        static_assert(noexcept(std::basic_string_view<char_type>(str(), size())));
        return std::basic_string_view<char_type>(str(), size());
    }
```

and as far as I can tell those functions are not inlined nor executed in compile time, so at least 3 function calls and static asserts. Why to slow down compilation and runtime if you can just export all those implemented functions (even if it's very tiny slowdown).

File: alternator/serialization.cc
Comment (nuivall) [reply alternator/serialization.cc:62]: Yes I also think it is. I'd leave the fixme as the code can change and suddenly it won't be correct, `from_json_visitor` seems to be a generic code.

**Comment by nuivall** [comment]
Thanks for the next review round! I pushed all requested changes.

File: schema.cc
Comment (nyh) [reply schema.cc]: The document introducing std::string_view, https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3921.html, mentions that "both gcc and clang optimize strlen("Literal") into a constant, making the simple, safe code as efficient as the template."

File: utils/rjson.hh
Comment (nyh) [reply utils/rjson.hh:192]: C++ is a very slow to compile language, because so much happens at compile time. We can't really fight it, and the compile-time overhead of implicit conversion of std::string or "string literal" to std::string_view  is probably negligible compared to all other compile-time overheads.

One of the reasons std::string_view was introduced was to replace all those uses of `std::string&`, and now you want to bring it back. Bringing back `std::string&` and in our case also `sstring&` is not such a disaster, but I still don't understand where is the benefit :-( It seems to me that it made many of the calls uglier, and a few less efficient. I don't see which of the calls improved its run-time performance by using the `sstring&` overload instead of `std::string_view`.

File: alternator/conditions.cc
Comment (nyh) [reply alternator/conditions.cc]: Yes. Maybe in the future (or already today, I didn't verify) this function may throw some other error (like out of memory) which doesn't mean it's an unparsable string, and shouldn't be swallowed as we do today?

If you're sure that base64_begins_with is guaranteed to only throw exceptions in the parse-error case then my suggestion isnt necessary?

File: test/alternator/test_manual_requests.py
Comment (nyh) [reply test/alternator/test_manual_requests.py:301]: Yes, it's not critical to have this test and I don't know if we'll ever bother to fix it even if we do. So your call.

File: schema.cc
Comment (nuivall) [reply schema.cc]: true, it does copy, I didn't follow the code to the deepest call

File: utils/rjson.hh
Comment (nuivall) [reply utils/rjson.hh:192]: maybe string -> string_view is efficiently  optimized (not sure) but sstring -> string_view calls https://github.com/scylladb/seastar/blob/master/include/seastar/core/sstring.hh#L579 which looks less efficient than just choosing the right overload

in which cases using those extra overloads would be slower, to me seems like callers have full control which overload they want?

to be honest I'd not add those functions if they would not be implemented, when I saw odd stuff like `rjson::from_string(sstring(_collection_name.begin(), _collection_name.end())));` I just though they were accidentally not exposed. If most pieces of the code would use string_view that'd also be a noble quest to try to unify. But to me looks like this attempt (if it was conscious) was slightly artificial as most or many usages are just sstring or string.

**Comment by nuivall** [comment]
@nyh I separated the commit you questioned to https://github.com/scylladb/scylladb/pull/12535 as it's not touching the original issue this patch tackles. I'll decide later what to do with it, either better prove usefulness or drop.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
@nuivall I wanted to merge this PR, but I see you pushed a new version two hours ago and the CI hasn't completed yet, so I'll merge it when it does. Thanks.

**Pull request closed without merging**
